src/dividat-driver/senso/websocket.go

```diff
@@ -95,15 +110,21 @@ func (command *Command) UnmarshalJSON(data []byte) error {
 // Message that can be sent to Play
 type Message struct {
 	*Status
-
-	Discovered *zeroconf.ServiceEntry
+	Discovered            *zeroconf.ServiceEntry
+	FirmwareUpdateMessage *FirmwareUpdateMessage
 }
 
 // Status is a message containing status information
 type Status struct {
 	Address *string
 }
 
+type FirmwareUpdateMessage struct {
+	FirmwareUpdateProgress *string
+	FirmwareUpdateSuccess  *string
+	FirmwareUpdateFailure  *string
+}
+
```

Comment:
I'm still wondering about whether to simplify this into something like:

```go
type FirmwareUpdateMessage struct {
	Type    string
	Message string
}
```

It could simplify code both here and the client. A potential downside of merging the three messages is that it would make it more difficult to evolve them independently, eg. add some fields to some of them but not the others.

Reply:
I quickly tried the alternative I came to the conclusion that the representation is fine as it is, but I'd simplify json serialization to remove a bunch of code:

```diff
diff --git a/src/dividat-driver/senso/websocket.go b/src/dividat-driver/senso/websocket.go
index ce19d83..763e18d 100644
--- a/src/dividat-driver/senso/websocket.go
+++ b/src/dividat-driver/senso/websocket.go
@@ -110,8 +110,8 @@ func (command *Command) UnmarshalJSON(data []byte) error {
 // Message that can be sent to Play
 type Message struct {
 	*Status
-	Discovered            *zeroconf.ServiceEntry
-	FirmwareUpdateMessage *FirmwareUpdateMessage
+	Discovered             *zeroconf.ServiceEntry
+	FirmwareUpdateResponse *FirmwareUpdateMessage
 }
 
 // Status is a message containing status information
@@ -119,10 +119,10 @@ type Status struct {
 	Address *string
 }
 
+// Message to update client about firmware update progress
 type FirmwareUpdateMessage struct {
-	FirmwareUpdateProgress *string
-	FirmwareUpdateSuccess  *string
-	FirmwareUpdateFailure  *string
+	Type    string `json:"type"`
+	Message string `json:"message"`
 }
 
 // MarshalJSON ipmlements JSON encoder for messages
@@ -147,34 +147,8 @@ func (message *Message) MarshalJSON() ([]byte, error) {
 			IP:           append(message.Discovered.AddrIPv4, message.Discovered.AddrIPv6...),
 		})
 
-	} else if message.FirmwareUpdateMessage != nil {
-		fwUpdate := struct {
-			Type    string `json:"type"`
-			Message string `json:"message"`
-		}{}
-
-		firmwareUpdateMessage := *message.FirmwareUpdateMessage
-
-		if firmwareUpdateMessage.FirmwareUpdateProgress != nil {
-
-			fwUpdate.Type = "FirmwareUpdateProgress"
-			fwUpdate.Message = *firmwareUpdateMessage.FirmwareUpdateProgress
-
-		} else if firmwareUpdateMessage.FirmwareUpdateFailure != nil {
-
-			fwUpdate.Type = "FirmwareUpdateFailure"
-			fwUpdate.Message = *firmwareUpdateMessage.FirmwareUpdateFailure
-
-		} else if firmwareUpdateMessage.FirmwareUpdateSuccess != nil {
-
-			fwUpdate.Type = "FirmwareUpdateSuccess"
-			fwUpdate.Message = *firmwareUpdateMessage.FirmwareUpdateSuccess
-
-		} else {
-			return nil, errors.New("could not marshal firmware update message")
-		}
-
-		return json.Marshal(fwUpdate)
+	} else if message.FirmwareUpdateResponse != nil {
+		return json.Marshal(message.FirmwareUpdateResponse)
 	}
 
 	return nil, errors.New("could not marshal message")
@@ -356,33 +330,21 @@ func (handle *Handle) dispatchCommand(ctx context.Context, log *logrus.Entry, co
 	} else if command.UpdateFirmware != nil {
 		go handle.ProcessFirmwareUpdateRequest(*command.UpdateFirmware, SendMsg{
 			progress: func(msg string) {
-				sendMessage(firmwareUpdateProgress(msg))
+				sendMessage(firmwareUpdateMessage("FirmwareUpdateProgress", msg))
 			},
 			failure: func(msg string) {
-				sendMessage(firmwareUpdateFailure(msg))
+				sendMessage(firmwareUpdateMessage("FirmwareUpdateFailure", msg))
 			},
 			success: func(msg string) {
-				sendMessage(firmwareUpdateSuccess(msg))
+				sendMessage(firmwareUpdateMessage("FirmwareUpdateSuccess", msg))
 			},
 		})
 	}
 	return nil
 }
 
-func firmwareUpdateSuccess(msg string) Message {
-	return firmwareUpdateMessage(FirmwareUpdateMessage{FirmwareUpdateSuccess: &msg})
-}
-
-func firmwareUpdateFailure(msg string) Message {
-	return firmwareUpdateMessage(FirmwareUpdateMessage{FirmwareUpdateFailure: &msg})
-}
-
-func firmwareUpdateProgress(msg string) Message {
-	return firmwareUpdateMessage(FirmwareUpdateMessage{FirmwareUpdateProgress: &msg})
-}
-
-func firmwareUpdateMessage(msg FirmwareUpdateMessage) Message {
-	return Message{FirmwareUpdateMessage: &msg}
+func firmwareUpdateMessage(responseType string, msg string) Message {
+	return Message{FirmwareUpdateResponse: &FirmwareUpdateMessage{Type: responseType, Message: msg}}
 }
 
 // rx_data_loop reads data from Senso and forwards it up the WebSocket

```

However this would introduce some inconsistency to how firmware update messages vs all the other messages are serialized, so I'd rather keep things as they are now. I think it would be worth refactoring, but it would be better to change all of the messages at once to follow the same pattern.

Reply:
I wonder whether the same pattern would work for the others, seeing that unlike with these three, each variant carries different fields?

Maybe type switches could be useful? https://go.dev/tour/methods/16

Reply:
I think something like this could work.

<details><summary>Diff</summary>
<p>

```diff
diff --git a/src/dividat-driver/senso/messages.go b/src/dividat-driver/senso/messages.go
new file mode 100644
index 0000000..2149687
--- /dev/null
+++ b/src/dividat-driver/senso/messages.go
@@ -0,0 +1,64 @@
+package senso
+
+import (
+	"encoding/json"
+	"net"
+
+	"github.com/libp2p/zeroconf/v2"
+)
+
+type Message struct {
+	*Status
+	*Discovered
+	*FirmwareUpdateResponse
+	Type string `json:"type"`
+}
+
+// Status
+
+type Status struct {
+	Address *string `json:"address"`
+}
+
+func (s *Status) MarshalJSON() ([]byte, error) {
+	return json.Marshal(Message{
+		Type:   "Status",
+		Status: s,
+	})
+}
+
+// Discovered
+
+type Discovered struct {
+	ServiceEntry *zeroconf.ServiceEntry `json:"service"`
+	IP           []net.IP               `json:"ip"`
+}
+
+func (d *Discovered) MarshalJSON() ([]byte, error) {
+	return json.Marshal(Message{
+		Type:       "Discovered",
+		Discovered: d,
+	})
+}
+
+// Firmware update messages
+
+type FirmwareUpdateResponseType string
+
+const (
+	FirmwareUpdateSuccess  FirmwareUpdateResponseType = "FirmwareUpdateSuccess"
+	FirmwareUpdateFailure  FirmwareUpdateResponseType = "FirmwareUpdateFailure"
+	FirmwareUpdateProgress FirmwareUpdateResponseType = "FirmwareUpdateProgress"
+)
+
+type FirmwareUpdateResponse struct {
+	Type    FirmwareUpdateResponseType
+	Message string `json:"message"`
+}
+
+func (r *FirmwareUpdateResponse) MarshalJSON() ([]byte, error) {
+	return json.Marshal(Message{
+		Type:                   string(r.Type),
+		FirmwareUpdateResponse: r,
+	})
+}
diff --git a/src/dividat-driver/senso/websocket.go b/src/dividat-driver/senso/websocket.go
index ce19d83..eb7ba53 100644
--- a/src/dividat-driver/senso/websocket.go
+++ b/src/dividat-driver/senso/websocket.go
@@ -4,13 +4,11 @@ import (
 	"context"
 	"encoding/json"
 	"errors"
-	"net"
 	"net/http"
 	"sync"
 	"time"
 
 	"github.com/gorilla/websocket"
-	"github.com/libp2p/zeroconf/v2"
 	"github.com/sirupsen/logrus"
 
 	"github.com/dividat/driver/src/dividat-driver/service"
@@ -107,80 +105,6 @@ func (command *Command) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
-// Message that can be sent to Play
-type Message struct {
-	*Status
-	Discovered            *zeroconf.ServiceEntry
-	FirmwareUpdateMessage *FirmwareUpdateMessage
-}
-
-// Status is a message containing status information
-type Status struct {
-	Address *string
-}
-
-type FirmwareUpdateMessage struct {
-	FirmwareUpdateProgress *string
-	FirmwareUpdateSuccess  *string
-	FirmwareUpdateFailure  *string
-}
-
-// MarshalJSON ipmlements JSON encoder for messages
-func (message *Message) MarshalJSON() ([]byte, error) {
-	if message.Status != nil {
-		return json.Marshal(&struct {
-			Type    string  `json:"type"`
-			Address *string `json:"address"`
-		}{
-			Type:    "Status",
-			Address: message.Status.Address,
-		})
-
-	} else if message.Discovered != nil {
-		return json.Marshal(&struct {
-			Type         string                 `json:"type"`
-			ServiceEntry *zeroconf.ServiceEntry `json:"service"`
-			IP           []net.IP               `json:"ip"`
-		}{
-			Type:         "Discovered",
-			ServiceEntry: message.Discovered,
-			IP:           append(message.Discovered.AddrIPv4, message.Discovered.AddrIPv6...),
-		})
-
-	} else if message.FirmwareUpdateMessage != nil {
-		fwUpdate := struct {
-			Type    string `json:"type"`
-			Message string `json:"message"`
-		}{}
-
-		firmwareUpdateMessage := *message.FirmwareUpdateMessage
-
-		if firmwareUpdateMessage.FirmwareUpdateProgress != nil {
-
-			fwUpdate.Type = "FirmwareUpdateProgress"
-			fwUpdate.Message = *firmwareUpdateMessage.FirmwareUpdateProgress
-
-		} else if firmwareUpdateMessage.FirmwareUpdateFailure != nil {
-
-			fwUpdate.Type = "FirmwareUpdateFailure"
-			fwUpdate.Message = *firmwareUpdateMessage.FirmwareUpdateFailure
-
-		} else if firmwareUpdateMessage.FirmwareUpdateSuccess != nil {
-
-			fwUpdate.Type = "FirmwareUpdateSuccess"
-			fwUpdate.Message = *firmwareUpdateMessage.FirmwareUpdateSuccess
-
-		} else {
-			return nil, errors.New("could not marshal firmware update message")
-		}
-
-		return json.Marshal(fwUpdate)
-	}
-
-	return nil, errors.New("could not marshal message")
-
-}
-
 // Implement net/http Handler interface
 func (handle *Handle) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
@@ -222,7 +146,7 @@ func (handle *Handle) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 
 	// send messgae up the WebSocket
-	sendMessage := func(message Message) error {
+	sendMessage := func(message json.Marshaler) error {
 		writeMutex.Lock()
 		conn.SetWriteDeadline(time.Now().Add(50 * time.Millisecond))
 		err := conn.WriteJSON(&message)
@@ -307,15 +231,13 @@ func (handle *Handle) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 // HELPERS
 
 // dispatchCommand handles incomming commands and sends responses back up the WebSocket
-func (handle *Handle) dispatchCommand(ctx context.Context, log *logrus.Entry, command Command, sendMessage func(Message) error) error {
+func (handle *Handle) dispatchCommand(ctx context.Context, log *logrus.Entry, command Command, sendMessage func(json.Marshaler) error) error {
 
 	if command.GetStatus != nil {
 
-		var message Message
-
-		message.Status = &Status{Address: handle.Address}
-
-		err := sendMessage(message)
+		err := sendMessage(&Status{
+			Address: handle.Address,
+		})
 
 		if err != nil {
 			return err
@@ -339,10 +261,11 @@ func (handle *Handle) dispatchCommand(ctx context.Context, log *logrus.Entry, co
 			for entry := range entries {
 				log.WithField("service", entry).Debug("Discovered service.")
 
-				var message Message
-				message.Discovered = &entry.ServiceEntry
-
-				err := sendMessage(message)
+				serviceEntry := &entry.ServiceEntry
+				err := sendMessage(&Discovered{
+					ServiceEntry: serviceEntry,
+					IP:           append(serviceEntry.AddrIPv4, serviceEntry.AddrIPv6...),
+				})
 				if err != nil {
 					return
 				}
@@ -356,35 +279,19 @@ func (handle *Handle) dispatchCommand(ctx context.Context, log *logrus.Entry, co
 	} else if command.UpdateFirmware != nil {
 		go handle.ProcessFirmwareUpdateRequest(*command.UpdateFirmware, SendMsg{
 			progress: func(msg string) {
-				sendMessage(firmwareUpdateProgress(msg))
+				sendMessage(&FirmwareUpdateResponse{Type: FirmwareUpdateProgress, Message: msg})
 			},
 			failure: func(msg string) {
-				sendMessage(firmwareUpdateFailure(msg))
+				sendMessage(&FirmwareUpdateResponse{Type: FirmwareUpdateFailure, Message: msg})
 			},
 			success: func(msg string) {
-				sendMessage(firmwareUpdateSuccess(msg))
+				sendMessage(&FirmwareUpdateResponse{Type: FirmwareUpdateSuccess, Message: msg})
 			},
 		})
 	}
 	return nil
 }
 
-func firmwareUpdateSuccess(msg string) Message {
-	return firmwareUpdateMessage(FirmwareUpdateMessage{FirmwareUpdateSuccess: &msg})
-}
-
-func firmwareUpdateFailure(msg string) Message {
-	return firmwareUpdateMessage(FirmwareUpdateMessage{FirmwareUpdateFailure: &msg})
-}
-
-func firmwareUpdateProgress(msg string) Message {
-	return firmwareUpdateMessage(FirmwareUpdateMessage{FirmwareUpdateProgress: &msg})
-}
-
-func firmwareUpdateMessage(msg FirmwareUpdateMessage) Message {
-	return Message{FirmwareUpdateMessage: &msg}
-}
-
 // rx_data_loop reads data from Senso and forwards it up the WebSocket
 func rx_data_loop(ctx context.Context, rx chan interface{}, send func([]byte) error) {
 	var err error

```

</p>
</details>